receiver/hostmetricsreceiver/internal/scraper/cpuscraper/internal/metadata/generated_metrics.go

```diff
@@ -24,75 +24,190 @@ import (
 // Type is the component type name.
 const Type config.Type = "cpu"
 
-// MetricIntf is an interface to generically interact with generated metric.
-type MetricIntf interface {
+// DataPoint represents a data structure for pdata.DataPoint creation
+type DataPoint struct {
+	StartTimestamp pdata.Timestamp
+	Timestamp      pdata.Timestamp
+	AttributeCpu   string
+	AttributeState string
+	// We probably can have number type (int64 vs float64) defined in metadata, and keep only one field for Number value
+	IntVal    *int64
+	DoubleVal *float64
+	// To be generated for SummaryDataPoint and ExponentialHistogramDataPoint data types
+	// ValueCount int64
+	// ValueSum float64
+	// ...
+	Exemplars pdata.ExemplarSlice
+	Flags     pdata.MetricDataPointFlags
+}
+
+// MetricTemplate represents a template for pdata.Metric creation based on metadata and user configuration
+type MetricTemplate interface {
+	EnsureDataPointsCapacity(int)
+	AddDatapoint(DataPoint)
+	AppendToMetricSlice(pdata.MetricSlice)
+}
+
+type metricTemplate struct {
+	mb             metricBuilder
+	metric         pdata.Metric
+	startTimestamp pdata.Timestamp
+	timestamp      pdata.Timestamp
+}
+
+func (mt metricTemplate) EnsureDataPointsCapacity(cap int) {
+	if !mt.mb.Enabled() {
+		return
+	}
+	mt.metric.Sum().DataPoints().EnsureCapacity(cap)
+	// For other types:
+	// mt.metric.Gauge().DataPoints().EnsureCapacity(cap)
+	// mt.metric.Histogram().DataPoints().EnsureCapacity(cap)
+	// mt.metric.Summary().DataPoints().EnsureCapacity(cap)
+}
+
+// AddDatapoint
+func (mt metricTemplate) AddDatapoint(data DataPoint) {
+	if !mt.mb.Enabled() {
+		return
+	}
+
+	dp := mt.metric.Sum().DataPoints().AppendEmpty()
+	dp.SetStartTimestamp(data.StartTimestamp)
+	dp.SetTimestamp(data.Timestamp)
+	if data.AttributeCpu != "" {
+		dp.Attributes().UpsertString(L.Cpu, data.AttributeCpu)
+	}
+	if data.AttributeState != "" {
+		dp.Attributes().UpsertString(L.State, data.AttributeState)
+	}
+	if data.IntVal != nil {
+		dp.SetIntVal(*data.IntVal)
+	}
+	if data.DoubleVal != nil {
+		dp.SetDoubleVal(*data.DoubleVal)
+	}
+	data.Exemplars.CopyTo(dp.Exemplars())
+	dp.SetFlags(data.Flags)
+}
+
+func (mt metricTemplate) AppendToMetricSlice(metrics pdata.MetricSlice) {
+	if !mt.mb.Enabled() {
+		return
+	}
+	mt.metric.CopyTo(metrics.AppendEmpty())
+}
+
+var _ MetricTemplate = (*metricTemplate)(nil)
+
+// MetricBuilder is an interface to generate metric.
+type MetricBuilder interface {
 	Name() string
-	New() pdata.Metric
-	Init(metric pdata.Metric)
+	Enabled() bool
+	Description() string
+	Unit() string
+	DataType() pdata.MetricDataType
+	IsMonotonic() bool
+	Temporality() pdata.MetricAggregationTemporality
```

Comment:
I was thinking of providing an alternative "manual" way for metrics building just in case the prescribed way doesn't work for scraper authors for any reason. Maybe we can exclude them for now and see i there will be a need for them going forward